Fix for #17593

khwilliamson opened the pull request:
Currently smoking at https://git.io/JvdFo

khwilliamson committed:
regcomp.c: Look before leaping; check before alloc

This moves the allocation of a new regnode to just after the check that
the contents are ok, instead of just before.

hvds commented on regcomp.c:
```diff
@@ -11710,7 +11710,18 @@ S_reg(pTHX_ RExC_state_t *pRExC_state, I32 paren, I32 *flagp, U32 depth)
-                if ( paren == '-' ) {
+                if (paren == '-' || paren == '+') {
+                    if (UNLIKELY(I32_MAX - RExC_npar < num)) {
+                        goto bad_group; /* Avoid overflowing, die
```
Rather than add more gotos, I'd be inclined to leave the disposition explicit (and consider making the error message more accurate for this case). If you're worried about duplicating the error string in the binary, I suggest giving it a name.

khwilliamson replied:
Care to suggest some wording?
Also, what about negative numbers too far from zero.  The comment just above is asking should it be concerned about this from the grok return

hvds replied:
"Invalid reference to group", perhaps?

I don't see how we can overflow on the negative side: I think at worst we'd substract `-I32_MAX` from 0. It's worth making sure we have test coverage though: I32_MIN or less should give invalid, others should give nonexistent group.

Come to think of it, I'm not sure I understand the handling of `if (num < 1)` at all: `RExC_npar` is our current paren isn't it? So that won't change after a REQUIRE_PARENS_PASS. Unless I have completely the wrong end of the stick, only a '+' can give us a forward reference.

khwilliamson replied:
I have to think further about your later points in your comment, but a reason I didn't add a new error is because the current code doesn't error on a number that's above I32_MAX, it just sets it to the max value.  It seems to me that using an input number that is out of bounds should give the same error as the one you're proposing.  Agreed?

hvds replied:
> It seems to me that using an input number that is out of bounds should give the same error as the one you're proposing. Agreed?

Yes, agreed. And though LTA it would not be a disaster if they all gave "nonexistent group"; I'd just prefer to avoid adding gotos to achieve that unless we need to.

khwilliamson replied:
See updated PR

hvds replied:
Looks good to me apart from the one new comment.

khwilliamson replied:
Fixed your point, and pushed

khwilliamson committed:
regcomp.c: Don't duplicate text for an error msg

Instead, have a common definition.

khwilliamson committed:
regcomp.c: Rmv wrong comment

This comment speculated that perhaps we should be testing for underflow.
But this isn't a problem on any extant computer hardware

khwilliamson committed:
regcomp.c: Rmv unnecessary code

As spotted by Hugo van der Sanden, a barckwards reference to a capturing
group doesn't need to look forward instead.  If we don't have a
capturing group we already should have, looking forward won't magically
produce it.

khwilliamson committed:
regcomp.c: Avoid a segfault

This was resulting in C undefined behavior reported by asan.  Check that
won't overflow before doing the operation; then die instead of going
ahead anyway.

This fixes #17593

khwilliamson committed:
regcomp.c: Die on relative group number overflow

That this code was doing in the presence of an illegally large (or
small) relative capturing group number was to set it to about the
furthest away from zero it could get, and silently carry on, where it
likely overflowed a few lines down.

Instead die immediately with a proper message.

hvds commented on regcomp.c:
```diff
@@ -11703,8 +11703,14 @@ S_reg(pTHX_ RExC_state_t *pRExC_state, I32 paren, I32 *flagp, U32 depth)
+                    }
+                    else {  /* Overflow, or something like that.  Position
+                               beyond all digits for the error */
+                        while (isDIGIT(*RExC_parse))  {
```
I think this needs to be:
```
  while (RExC_parse < RExC_end && isDIGIT(*RExC_parse)) {
```
